nncf/experimental/post_training/api/engine.py

```diff
@@ -19,30 +19,76 @@
 
 ModelType = TypeVar('ModelType')
 TensorType = TypeVar('TensorType')
-
+MetricType = TypeVar('MetricType')
 
 class Engine(ABC):
     """
     The basic class aims to provide the interface to infer the model.
     """
+    # TODO (Nikita Malinin): Update class with the _get_sampler() method
 
     def __init__(self):
         self.model = None
+        self._sampler = None
+        self._data_loader = None
+        self._metrics = None
+
+    # TODO (Nikita Malinin): Add statistic aggregator object (per-backend)
+    
+    @property
+    def data_loader(self):
+        return self._data_loader
+
+    @data_loader.setter
+    def data_loader(self, data_loader):
+        self._data_loader = data_loader
+
+    @property
+    def metrics(self):
+        return self._metrics
+
+    @metrics.setter
+    def metrics(self, metrics):
+        self._metrics = metrics
+
+    @property
+    def sampler(self):
+        return self._sampler
+
+    @sampler.setter
+    def sampler(self, sampler):
+        self._sampler = sampler
 
     def set_model(self, model: ModelType) -> None:
         self.model = model
 
     def is_model_set(self) -> bool:
         return self.model is not None
 
-    def infer(self, _input: TensorType) -> Dict[str, TensorType]:
-        if not self.is_model_set():
-            raise RuntimeError('The {} tried to infer the model, while the model was not set.'.format(self.__class__))
-        return self._infer(_input)
+    def transform_output(self, outputs: TensorType) -> TensorType:
+        """ Processes model output data
+            :param outputs: list of output data
+            :return outputs: list of the output data in an order expected by the accuracy metric if any is used
+        """
+        return outputs
 
-    @abstractmethod
-    def _infer(self, _input: TensorType) -> Dict[str, TensorType]:
+    def transform_input(self, inputs: TensorType) -> TensorType:
+        """ Processes model input data based on the backend
+            :param inputs: list of input data
+            :return inputs: list of the input data
         """
-        Infer the model on the provided input.
-        Returns the model outputs and corresponding node names in the model.
+        return inputs
+
+    def compute_statistics(self, statistics_layout: Dict) -> Dict[str, TensorType]:
+        """ Performs model inference on specified dataset subset for statistics collection and input-based layers layout
```

Comment:
Should check whether the model is set

Reply:
This is verified in the `ONNXEngine` implementation. There are no ideas yet on which default validation implementation will not overload the API and worsen the UX.

Reply:
I think that the checking of the model status is the first essential step for all Engines

Reply:
> I think that the checking of the model status is the first essential step for all Engines

Any suggestions to do this without overloading the API and worsening the UX?

Reply:
The solution could be as it is done right now for `Engine`:

```
    def infer(self, _input: TensorType) -> Dict[str, TensorType]:
        if not self.is_model_set():
            raise RuntimeError('The {} tried to infer the model, while the model was not set.'.format(self.__class__))
        return self._infer(_input)

    @abstractmethod
    def _infer(self, _input: TensorType) -> Dict[str, TensorType]:
        """
        Infer the model on the provided input.
        Returns the model outputs and corresponding node names in the model.
        """
```

Reply:
In the case of `Engine` re-implementation, the user should implement a private method?

Reply:
I see no problem here

Reply:
These checks have been moved to `calculate_statistics()` and `calculate_metrics()` respectively.